x/gasless/keeper/fee_helper.go

```diff
@@ -0,0 +1,241 @@
+	if fee.Amount.GT(gasTank.MaxFeeUsagePerTx) {
+		return gasTank, false, sdkerrors.Wrapf(types.ErrorFeeConsumptionFailure, "fee amount more than allowed limit")
+	}
+
+	// insufficient reserve in the gas tank to fulfill the transaction fee
+	gasTankReserveBalance := k.GetGasTankReserveBalance(ctx, gasTank)
+	if gasTankReserveBalance.IsLT(fee) {
+		return gasTank, false, sdkerrors.Wrapf(types.ErrorFeeConsumptionFailure, "funds insufficient in gas reserve tank")
+	}
+
+	consumptionFound := false
+	consumptionIndex := 0
+	for index, consumption := range consumer.Consumptions {
+		if consumption.GasTankId == gasTank.Id {
+			consumptionIndex = index
+			consumptionFound = true
+		}
+	}
+	// no need to check the consumption usage since there is no key available with given gas tank id
+	// i.e the consumer has never used this gas reserve before and the first time visitor for the given gas tank
+	if !consumptionFound {
+		return gasTank, true, nil
+	}
+
+	consumptionDetails := consumer.Consumptions[consumptionIndex]
+
+	// consumer is blocked by the gas tank
+	if consumptionDetails.IsBlocked {
+		return gasTank, false, sdkerrors.Wrapf(types.ErrorFeeConsumptionFailure, "blocked by gas tank")
+	}
+
+	// consumer exhausted the transaction count limit, hence not eligible with given gas tank
+	if consumptionDetails.TotalTxsMade >= consumptionDetails.TotalTxsAllowed {
+		return gasTank, false, sdkerrors.Wrapf(types.ErrorFeeConsumptionFailure, "exhausted tx limit")
+	}
+
+	// if total fees consumed by the consumer is more than or equal to the allowed consumption
+	// i.e consumer has exhausted its fee limit and hence is not eligible for the given tank
+	totalFeeConsumption := consumptionDetails.TotalFeesConsumed.Add(fee.Amount)
+	if totalFeeConsumption.GTE(consumptionDetails.TotalFeeConsumptionAllowed) {
+		return gasTank, false, sdkerrors.Wrapf(types.ErrorFeeConsumptionFailure, "exhausted total fee usage or pending fee limit insufficient for tx")
+	}
+
+	return gasTank, true, nil
+}
+
+func (k Keeper) GetFeeSource(ctx sdk.Context, sdkTx sdk.Tx, originalFeePayer sdk.AccAddress, fees sdk.Coins) sdk.AccAddress {
+	if len(sdkTx.GetMsgs()) > 1 {
+		k.EmitFeeConsumptionEvent(ctx, originalFeePayer, []uint64{}, []error{sdkerrors.Wrapf(types.ErrorFeeConsumptionFailure, "multiple messages")}, 0)
+		return originalFeePayer
+	}
+
+	// only one fee coin is supported, tx containing multiple coins as fees are not allowed.
+	if len(fees) != 1 {
+		k.EmitFeeConsumptionEvent(ctx, originalFeePayer, []uint64{}, []error{sdkerrors.Wrapf(types.ErrorFeeConsumptionFailure, "asked fee != 1")}, 0)
+		return originalFeePayer
+	}
+
+	fee := fees[0]
+
+	msg := sdkTx.GetMsgs()[0]
+	msgTypeURL := sdk.MsgTypeURL(msg)
+
+	isContract := false
+	var contractAddress string
+
+	executeContractMessage, ok := msg.(*wasmtypes.MsgExecuteContract)
+	if ok {
+		isContract = true
+		contractAddress = executeContractMessage.GetContract()
+	}
```

Comment:
not generic over other VMs

Reply:
refactored, and moved to `x/gasless/keeper/usage_identifiers.go`, where other VMs can be added